content/docs/reference/reproducibility.md

```diff
@@ -0,0 +1,41 @@
++++
+title="Reproducible Builds"
+weight=9
++++
+
+## Summary
+The Cloud Native Buildpacks project aims to create "Reproducible Builds" of container images. For image creation commands (`create-builder`, `package-buildpackag`, `build`) `pack` aims to create in a reproducible fashion. "Reproducible" is hard to define but we'll do so by example:
+
+---
+Running `pack build` produce a container image with the same image ID (*local* case)
+
+**Given**:
+- A workspace directory containing the same source code
+- A builder image with a *given*  
+- One or more buildpacks that produce identical layers given their input*
+
+---
+Running `pack build --publish` produce a container image with the same image digest (*remote* case)
+
+**Given**:
+- A workspace directory containing the same source code
+- A builder image with a *given*  
+- One or more buildpacks that produce identical layers given their input
+
+### Consequences and Caveats
+
+We achieve reproducible builds by "zeroing" various timestamps of the layers that `pack` creates. When images are inspected (via something like `docker inspect`) they may have confusing creation times:
+
+```bash
+REPOSITORY                                   TAG                 IMAGE ID            CREATED             SIZE
+cnbs/sample-builder                          <none>              def52b23918d        40 years ago        234MB
+sample-kotlin-app                            alpine              45dc2d2681a1        40 years ago        18.9MB
+```
+
+All that said, the CNB lifecycle cannot fix non-reproducible buildpack layer file contents. This means that the underlying buildpack and language ecosystem have to implement reproducible output (for example `go` binaries are reproducible by default).
+
+A local and remote build will not produce the same image digest because:
```

Comment:
This is very interesting. If true, I think we should dig into it deeper. Why does the local image need to be an image ID instead of a reference? References can existing in the daemon. See `docker tag` command.

Reply:
I think the key here is that local daemon stored images won't have a digest until pushed (this is because of docker):
```bash
○ → docker inspect cnbs/sample-builder:alpine | jq --raw-output '.[0].RepoDigests[]'
cnbs/sample-builder@sha256:9e3cfea3f90fb4fbbe855a2cc9ce505087ae10d6805cfcb44bd67a4b72628641

  |2.6.3| NY-Floater-15565 in ~/workspace
○ → docker inspect cnb-test | jq --raw-output '.[0].RepoDigests[]'
Error: No such object: cnb-test
jq: error (at <stdin>:1): Cannot iterate over null (null)
```

Reply:
I did confirm that the `runImage.reference` does change between whether you run `pack build` or `pack build ... --publish`. If we could calculate the image digest of a local image I'm curious why we don't. This is now totally irrelevant to this issue. Just something that was brought to light so thank you.